FinniversKit/Sources/Components/KeyValueGridView/KeyValueGridView.swift

```diff
@@ -119,6 +164,230 @@ public class KeyValueGridView: UIView {
+            origin.y = buttonFrame.minY - tooltipHeight
+        case .bottom:
+            origin.x = buttonFrame.midX - (tooltipWidth / 2)
+            origin.y = buttonFrame.maxY
+        case .leading:
+            origin.x = buttonFrame.minX - tooltipWidth
+            origin.y = buttonFrame.midY - (tooltipHeight / 2)
+        case .trailing:
+            origin.x = buttonFrame.maxX
+            origin.y = buttonFrame.midY - (tooltipHeight / 2)
+        }
+
+        return origin
+    }
+
+    /// Ensures the tooltip doesn’t overflow outside the screen bounds
+    private func clampToScreenEdges(frame: CGRect, in screenBounds: CGRect, margin: CGFloat) -> CGRect {
+        var result = frame
+
+        // Left clamp
+        if result.minX < margin { result.origin.x = margin }
+        // Right clamp
+        if result.maxX > screenBounds.maxX - margin { result.origin.x = screenBounds.maxX - margin - result.width }
+        // Top clamp
+        if result.minY < margin { result.origin.y = margin }
+        // Bottom clamp
+        if result.maxY > screenBounds.maxY - margin { result.origin.y = screenBounds.maxY - margin - result.height }
+
+        return result
+    }
+
+    @objc private func handleTooltipTap(_ gesture: UITapGestureRecognizer) {
+        dismissTooltip()
+    }
+
+    private func dismissTooltip() {
+        activeTooltipView?.removeFromSuperview()
+        UIAccessibility.post(notification: .layoutChanged, argument: activeTooltipSource)
+        activeTooltipView = nil
+        activeTooltipSource = nil
+        initialTooltipSourceFrame = nil
+        stopTooltipDisplayLink()
+    }
+
+    func dismissAllTooltips() {
+        dismissTooltip()
+    }
+
+    // MARK: - Logic for dismissing tooltip when scrolling
+    private func startTooltipDisplayLink() {
+        stopTooltipDisplayLink()
+        tooltipDisplayLink = CADisplayLink(target: self, selector: #selector(checkTooltipSourceFrame))
+        tooltipDisplayLink?.add(to: .main, forMode: .default)
+    }
+
+    private func stopTooltipDisplayLink() {
+        tooltipDisplayLink?.invalidate()
+        tooltipDisplayLink = nil
+    }
+
+    @objc private func checkTooltipSourceFrame() {
+        guard let window = self.window,
+              let sourceView = activeTooltipSource,
+              let initialFrame = initialTooltipSourceFrame else { return }
+
+        let currentFrame = sourceView.convert(sourceView.bounds, to: window)
+        let threshold: CGFloat = 10.0
+        if abs(currentFrame.midX - initialFrame.midX) > threshold ||
+            abs(currentFrame.midY - initialFrame.midY) > threshold {
+            dismissTooltip()
+        }
+    }
+
+    private func findScrollView() -> UIScrollView? {
+        var view: UIView? = self
+        while view != nil {
+            if let scrollView = view as? UIScrollView {
+                return scrollView
+            }
+            view = view?.superview
+        }
+        return nil
```

Comment:
Not a needed change at all, but we could use [`sequence(first:next:)`](https://developer.apple.com/documentation/swift/sequence(first:next:)) here instead to avoid the control variables.

```swift
for view in sequence(first: self, next: { $0.superview }) {
    if let scrollView = view as? UIScrollView {
        return scrollView
    }
}
return nil
```